xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,201 @@
+	if c.clusterUpdate.ClusterType != xdsclient.ClusterTypeAggregate {
+		print("Returning the cluster update for service (base node):", c.clusterUpdate.ServiceName)
+		return []xdsclient.ClusterUpdate{c.clusterUpdate}, nil
+	}
+
+	// If an aggregate construct a list by recursively calling down to all of it's children.
+	var childrenUpdates []xdsclient.ClusterUpdate
+	for _, child := range c.children {
+		childUpdateList, err := child.constructClusterUpdate()
+		if err != nil {
+			return nil, err
+		}
+		childrenUpdates = append(childrenUpdates, childUpdateList...)
+	}
+	print("Returning the update list for an aggregate cluster")
+	return childrenUpdates, nil
+}
+
+// handleResp handles a xds response for a particular cluster. This function also
+// handles any logic with regards to any child state that may have changed.
+// Problem statement: perhaps this logic is wrong
+func (c *clusterNode) handleResp(clusterUpdate xdsclient.ClusterUpdate, err error) {
+	c.clusterHandler.clusterMutex.Lock()
+	defer c.clusterHandler.clusterMutex.Unlock()
+	print("SERVICE NAME:", clusterUpdate.ServiceName)
+	if err != nil { // Write this error for run() to pick up in CDS LB policy.
+		c.clusterHandler.updateChannel<-clusterHandlerUpdate{chu: nil, err: err}
+		return
+	}
+	// This variable will determine whether there was a delta with regards to this clusterupdate. If there was, at the end
+	// of the response ping ClusterHandler to send CDS Policy a new list of ClusterUpdates.
+	var delta bool
+	// If this is the first update to the Cluster Node, ping ClusterHandler at the end of method to try and build a config.
+	if !c.receivedUpdate {
+		delta = true
+	}
+
+	c.receivedUpdate = true
+	c.clusterUpdate = clusterUpdate
+
+	// This map will be empty if the cluster update specifies cluster is an EDS or LogicalDNS cluster, as will have no children.
+	newChildren := make(map[string]struct{})
+	if clusterUpdate.ClusterType == xdsclient.ClusterTypeAggregate {
+		for _, childName := range clusterUpdate.PrioritizedClusterNames {
+			print("New children added: ", childName)
+			newChildren[childName] = struct{}{}
+		}
+	}
+
+	for _, child := range c.children {
+		// If the child is still present in the update, then there is nothing to do for that child name in the update.
+		if _, found := newChildren[child.clusterUpdate.ServiceName]; found {
+			print("New children deleted: ", child.clusterUpdate.ServiceName)
+			delete(newChildren, child.clusterUpdate.ServiceName)
+		} else { // If the child is no longer present in the update, that cluster can be deleted.
+			delta = true
+			child.delete()
+		}
+	}
+
+	// Whatever clusters are left over here from the update are all new children, so create CDS watches for those clusters.
+	for child, _ := range newChildren {
+		print("Creating cds watch in:", child)
+		delta = true
+		c.children = append(c.children, createClusterNode(child, c.clusterHandler.xdsClient, c.clusterHandler))
```

Comment:
The children list is not a set, but a list sorted by priority.
Using a map [shuffles the items](https://blog.golang.org/maps#TOC_7.). But the order is important.

Reply:
100% right, thanks for catching this. I found a solution that replaces this code block. It's to use the initial cluster list as the ordering (as even the list of children currently might be reordered), and either just pull the pointer from the current children or construct a new ClusterNode.

Reply:
Solution is implemented.